PolyEngine/Engine/Src/ConfigBase.cpp

```diff
@@ -1,18 +1,86 @@
 #include "EnginePCH.hpp"
 #include "ConfigBase.hpp"
 
+#include <RTTISerialization.hpp>
+#include <sstream>
+#include <rapidjson/stringbuffer.h>
+#include <rapidjson/prettywriter.h>
+
+using namespace Poly;
+
 RTTI_DEFINE_TYPE(Poly::ConfigBase)
 
-Poly::ConfigBase::ConfigBase(const String& displayName, eConfigLocation location)
+ConfigBase::ConfigBase(const String& displayName, eConfigLocation location)
 	: DisplayName(displayName), Location(location)
 {
-	FileName = GetTypeInfo().GetTypeName();
 }
 
-void Poly::ConfigBase::Save()
+void ConfigBase::Save()
+{
+	rapidjson::Document DOMObject; // UTF8 by default
+	RTTI::SerializeObject(this, DisplayName, DOMObject);
+	
+	// Save to pretty string
+	rapidjson::StringBuffer buffer;
+	buffer.Clear();
+	rapidjson::PrettyWriter<rapidjson::StringBuffer> writer(buffer);
+	DOMObject.Accept(writer);
+	
+	gConsole.LogDebug("{} json:\n{}", DisplayName, buffer.GetString());
+	switch (Location)
+	{
+	case Poly::eConfigLocation::ENGINE:
+		SaveTextFileRelative(eResourceSource::ENGINE, GetFileName(), String(buffer.GetString()));
+		break;
+	case Poly::eConfigLocation::GAME:
+		SaveTextFileRelative(eResourceSource::GAME, GetFileName(), String(buffer.GetString()));
+		break;
+	case Poly::eConfigLocation::LOCAL:
+		SaveTextFile(GetFileName(), String(buffer.GetString()));
+		break;
+	default:
+		ASSERTE(false, "Invalid location");
+	}
+}
+
+void ConfigBase::Load()
 {
+	String json;
+	try
+	{
+		switch (Location)
+		{
+		case Poly::eConfigLocation::ENGINE:
+			json = LoadTextFileRelative(eResourceSource::ENGINE, GetFileName());
+			break;
+		case Poly::eConfigLocation::GAME:
+			json = LoadTextFileRelative(eResourceSource::GAME, GetFileName());
+			break;
+		case Poly::eConfigLocation::LOCAL:
+			json = LoadTextFile(GetFileName());
+			break;
+		default:
+			ASSERTE(false, "Invalid location");
+		}
+	} 
+	catch (std::exception)
+	{
+		gConsole.LogWarning("No configuration file found for {}. Using default values.", DisplayName);
+		Save(); // Create file
+		return;
+	}
+		
+	rapidjson::Document DOMObject;
+	DOMObject.Parse(json.GetCStr());
+	RTTI::DeserializeObject(this, DisplayName, DOMObject);
+
+	// For now, ensure newest state of config file.
+	Save(); 
 }
 
-void Poly::ConfigBase::Load()
+const String& ConfigBase::GetFileName() const
 {
+	if (FileName.GetLength() == 0)
+		FileName = String("Configs/") + String(GetTypeInfo().GetTypeName()) + String(".json");
+	return FileName;
```

Comment:
Why not make _FileName_ a static variable inside of this function, since it is not used anywhere outside of this getter anyway?
Instead of making the hack with mutable in _.hpp_ file?

Reply:
Because FileName for different configs will be different.

Reply:
I wonder if using a _curiously recurring template pattern_ here would be an overkill...
i.e. `DebugConfig : ConfigBase<DebugConfig>` which would yield a unique static variable for each derived config.

Reply:
This is totally unnecessary. Variable works just fine.

Reply:
Yeah, I was just throwing ideas around for the sake of it. Caching `mutable` variable seems fine to me (simpler with the same performance characteristic).